File: 2023/day4/main.py
from typing import List


def read_input(file: str) -> List[str]:
    with open(file, mode="r") as f:
        out = f.readlines()
    return [l.replace("\n", "") for l in out]


def card_matches(card: str) -> int:
    _, numbers = card.split(":")
    win, drawn = numbers.split("|")
    win = win.strip().split(" ")
    drawn = drawn.strip().split(" ")
    win = {int(x) for x in win if x != ""}
    drawn = {int(x) for x in drawn if x != ""}
    return len(win.intersection(drawn))


def main() -> None:
    input = read_input("input.txt")
    matches = [card_matches(c) for c in input]

    part1 = sum(2 ** (n - 1) for n in matches if n > 0)
    print(f"Part 1: {part1}")

    n_cards = len(input)
    cards = [1 for _ in range(0, n_cards)]
    for i, m in enumerate(matches):
        n = cards[i]
        if m == 0:
            continue
        for j in range(i + 1, min(i + m + 1, n_cards)):
            cards[j] = cards[j] + n

    part2 = sum(cards)
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()

File: 2023/day4/src/main.rs
use std::fs::read_to_string;
use::std::cmp::min;

fn read_input(filename: &str) -> Vec<String> {
    read_to_string(filename)
        .unwrap()
        .lines()
        .map(String::from)
        .collect()
}

fn number_matches(card: &str) -> usize {
    let mut values = card.split(':').nth(1).unwrap().split('|');

    let winning: Vec<usize> = values
        .next()
        .expect("Should be winning values")
        .split(' ')
        .flat_map(|x| x.parse::<usize>())
        .collect();

    values
        .next()
        .expect("Should be drawn values")
        .split(' ')
        .flat_map(|x| x.parse::<usize>())
        .filter(|v| winning.contains(v))
        .count()
}

fn main() {
    let input = read_input("input.txt");
    let matches: Vec<usize> = input
        .iter()
        .map(|s| number_matches(s))
        .collect();
    
    let part1: usize = matches
        .iter()
        .filter(|&n| *n != 0)
        .map(|n| 1 << (n - 1))
        .sum();

    println!("Part 1: {}", part1);

    let n_cards = matches.len();
    let mut cards = vec![1; n_cards];
    for (i, m) in matches.iter().enumerate().filter(|(_, &m)| m > 0) {
        let end = min(i + m + 1, n_cards);
        for j in (i + 1)..end {
            let n = cards[i]; 
            cards[j] += n;
        }
    }

    let part2: usize = cards.iter().sum();

    println!("Part 2: {}", part2);
}
